### PR TITLE
shortint/fix: take modular inputs on operations 

### PR DESCRIPTION
**Pull request by mayeul-zama**

### CODE REVIEW

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs:63]: Yes, indeed, when overflow checks are disabled and the modulus is a power of 2, it's correct

File: tfhe/src/shortint/engine/server_side/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/mod.rs]: Done

File: tfhe/src/shortint/server_key/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs]: Added

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs]: Fixed

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs]: Use `wrapping_sub`, removed double modulus

File: tfhe/src/shortint/engine/server_side/div_mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/server_side/div_mod.rs:101]: not sure I like this style of lookup generation, it hides the parts where there are moduli especially for "low level" usage as we do here 🤔 

is there still a raw primitive without automated modulus management ?

Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/div_mod.rs:101]: Sure, `generate_lookup_table` still exists and is used in many places, especially in the integer API

Comment (IceTDrinker) [reply tfhe/src/shortint/engine/server_side/div_mod.rs:101]: ok, I'll check more thorougly but the changes seemed straightforward otherwise

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (IceTDrinker) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs]: when in doubt use .rem_euclid() for the "mathematical" modulo operator, only is an issue for signed integer types

**Comment by IceTDrinker** [comment]
please fix the commit messages not passing conformance before we trigger test suites

File: tfhe/src/shortint/engine/server_side/shift.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/shift.rs:23]: Yes
If this function was used to build the integer API, this would break things
But I also guess that this function is supposed to be used ignoring the carry bits

Comment (IceTDrinker) [reply tfhe/src/shortint/engine/server_side/shift.rs:23]: could be we'll know once you have pushed the updated commit respecting the conventional notaion for approval test runs

**Comment by mayeul-zama** [comment]
> please fix the commit messages not passing conformance before we trigger test suites

Actually I'm not sure the change in the last commit is correct which is why I separated it for someone to give their opinion on it

I can remove or squash the commit once I know whether the changes should be kept or not

**Comment by IceTDrinker** [comment]
are there some tests associated with those ?

**Comment by mayeul-zama** [comment]
> are there some tests associated with those ?

Yes, but I guess they are not part of the fast tests
I added asserts in the associated doc tests

File: tfhe/src/shortint/server_key/mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/mod.rs:383]: use the msg mod from parameters directly ?

File: tfhe/src/shortint/key_switching_key/mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/key_switching_key/mod.rs:105]: that's the cleartext as it's not encoded yet

File: tfhe/src/shortint/key_switching_key/mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/key_switching_key/mod.rs:165]: same thing

**Review by IceTDrinker — COMMENTED** [review]
A few last comments before an approval for test runs, I would like a second pair of eyes on this before we merge (though tests should be a pretty good indicator already)

File: tfhe/src/shortint/engine/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/mod.rs:117]: `shortint::key_switching_key::test::gen_multi_keys_test_truncate` fails on main with `attempt to multiply with overflow` (with overflow checks enabled)

But it's because of a bug which I'll push a fix for
I'll remove this wrapping_mul and add an assert `f_eval < message * carry`

I think we should enable overflow checks on CI tests

Comment (IceTDrinker) [reply tfhe/src/shortint/engine/mod.rs:117]: yeah for the overflow checks it's in the backlog

what's the bug ?

but I think the code itself should do a checked mul so that we can warn/crash on overflow here as it means the table is likely broken

File: tfhe/src/shortint/server_key/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs]: Done

File: tfhe/src/shortint/key_switching_key/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/key_switching_key/mod.rs]: Done

File: tfhe/src/shortint/key_switching_key/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/key_switching_key/mod.rs]: Done

File: tfhe/src/shortint/engine/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/mod.rs]: Yes, the assert should be equivalent to a checked mul

File: tfhe/src/shortint/server_key/tests/shortint.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/tests/shortint.rs:535]: was this test failing ?
otherwise I agree that this change could destroy the padding bit

File: tfhe/src/shortint/key_switching_key/mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/key_switching_key/mod.rs:130]: does this make sense functionally ? I'm not sure about what the casting was doing, if we are trashing the padding bit we may want to check what's going on

File: tfhe/src/shortint/engine/client_side.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/client_side.rs:307]: so that's an interesting idea but I would probably see something like an Ok(val) if padding bit is 0 and an Err(val) with a nice error formatting for the Err variant

I don't know if asserting here is a good idea, decrypting is not failing per se

File: tfhe/src/shortint/engine/mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/mod.rs:119]: so same thing having a bit on the padding bit is not illegal per se, but we may want to return the information as an Enum/Result down the line as this may lead to invalid computations later, don't know what the best here would be

@tmontaigu if you have an opinion on this ?

also I wonder if it may not break some stuff like Parmesan by Jakub which relied on some tricks to have negacyclic functions

**Review by IceTDrinker — COMMENTED** [review]
A few comments

Generally speaking when I'm making comments/asking questions I'm not always expecting a fix in the commits in the next iteration it's more to open the conversation, especially when adding asserts for some behavior as they may require a bit more of planning on our end :) (talking about the decrypt assert and the TLU generation)

File: tfhe/src/shortint/server_key/tests/shortint.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/tests/shortint.rs:535]: With the new `assert!(f_eval < modulus_sup as u64)` the `shortint_generate_lookup_table` test was failing with all the parameters I think

File: tfhe/src/high_level_api/shortints/tests.rs
Comment (tmontaigu) [inline tfhe/src/high_level_api/shortints/tests.rs:176]: This should not be needed here, the modulus should be applied internally, not at the user code level

Comment (mayeul-zama) [reply tfhe/src/high_level_api/shortints/tests.rs:176]: You mean `map` and `apply` should apply the modulus on the output of the closure itself?

If so, would you do the modulus in `GenericShortIntServerKey::bootstrap_(inplace)_with`?

Should the modulo be over the message space, the message and the carry space or the message space, carry space and padding bit? It seems to me that it depends on the user. We could have a function for each or leave it to the caller as it is now

Comment (tmontaigu) [reply tfhe/src/high_level_api/shortints/tests.rs:176]: To me in the HLP API, we expose things like FheUint2, FheUint3 where the modulus is 2**2, 2**3
the idea is that the carry system is hidden, so users only see the 2/3 bits of message

Comment (mayeul-zama) [reply tfhe/src/high_level_api/shortints/tests.rs:176]: Makes sense, will fix

Comment (mayeul-zama) [reply tfhe/src/high_level_api/shortints/tests.rs]: Done

Comment (tmontaigu) [reply tfhe/src/high_level_api/shortints/tests.rs]: Also the modulus used in the assert is wrong : it does `%2` but it should be "%4" 

(The 'bug' was there before your modifications)

File: tfhe/src/shortint/engine/mod.rs
Comment (tmontaigu) [reply tfhe/src/shortint/engine/mod.rs:119]: There is a place in the integers comparision where we do a PBS on values where the bit of padding could be 1

File: tfhe/src/shortint/engine/server_side/mod.rs
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/mod.rs:303]: I wonder, if this applies modulus to the input, would it be coherent for it to also apply the modulus on the output ?

Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/mod.rs:303]: I guess if we want to ignore the carry bits of the input, we also want to ignore the ones of the output
So IMO it makes sense to apply the same modulus on the ouput here

Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/mod.rs]: Done

File: tfhe/src/shortint/server_key/mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/mod.rs:370]: this is most likely a wrong usage of ^, this should be a power instead what's the status on https://github.com/zama-ai/tfhe-rs/pull/638 @tmontaigu ?

File: tfhe/src/shortint/engine/server_side/shift.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/server_side/shift.rs:74]: this changes the modulus used here and mismatches the one in the call to after_left_shift below

File: tfhe/src/high_level_api/shortints/server_key.rs
Comment (IceTDrinker) [inline tfhe/src/high_level_api/shortints/server_key.rs:434]: should this be the message_modulus or the max modulus including the carry ? as long as we are not trashing the padding bit ? same bellow

File: tfhe/src/shortint/engine/server_side/mod.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/server_side/mod.rs:304]: I think I disagree with that % modulus at the end, someone may want to have bits in the carries even if only applying a LUT on the message part of x

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/server_side/scalar_sub.rs:64]: is this still correct when x is replaced by x % modulus ?

File: tfhe/src/shortint/server_key/mod.rs
Comment (tmontaigu) [reply tfhe/src/shortint/server_key/mod.rs:370]: Its ready to be reviewed

File: tfhe/src/shortint/server_key/mod.rs
Comment (tmontaigu) [inline tfhe/src/shortint/server_key/mod.rs:412]: x^2 in the lut means xor(x, 2) not power so this might confuse readers

File: tfhe/src/high_level_api/shortints/server_key.rs
Comment (tmontaigu) [reply tfhe/src/high_level_api/shortints/server_key.rs:434]: I think it might be better to have it include the carry space also since using the carry space is inherent to the proper use of shortint :thinking:

File: tfhe/src/shortint/engine/server_side/shift.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/shift.rs:74]: Do we support applying LUT with `sk.message_modulus != ct.message_modulus`?

File: tfhe/src/high_level_api/shortints/server_key.rs
Comment (mayeul-zama) [reply tfhe/src/high_level_api/shortints/server_key.rs:434]: I guess it's moslty an interpretation of what users should expect from this function

But I agree, I could make more sense to include the carry modulus

Comment (IceTDrinker) [reply tfhe/src/high_level_api/shortints/server_key.rs:434]: I guess an unchecked_generated_lut would be nice ? and maybe a smart_generate_lut which would give the msg and carry values basically would do

```rust
for plaintext in 0..plaintext_mod {
    let msg = plaintext % msg_mod;
    let carry = plaintext / msg_mod;
    let lut_val = user_provided_fn(msg, carry);
    // store in lut
}
```

this way you can chose if you are manipulating the message and/or the carry

File: tfhe/src/shortint/engine/server_side/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/mod.rs:304]: See https://github.com/zama-ai/tfhe-rs/pull/538#discussion_r1334440315

IMO, if someone wants to play with the input or output carry, they should use `generate_lookup_table` and handle the input and output carry themselves

Comment (IceTDrinker) [reply tfhe/src/shortint/engine/server_side/mod.rs:304]: that's a design question, what is the meaning of the various functions we have in there, could you create a quick list of those function and their current behavior ? this way we can discuss and then agree on what we want, as those are mostly used by us (given shortint is already fairly low level) it's mainly a QoL for us so we can chose what we prefer as implementers

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs:64]: Yes, I think so

But there may be a problem with wrapping if `ct_modulus % message_modulus != 0`
I think `x + msg_modulus - (scalar % msg_modulus)` is what we want. Does it make sense to you?

File: tfhe/src/shortint/server_key/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs:412]: Yes, I thought it was a pow :upside_down_face:

Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs:412]: Should I change that to `x * x` ?

Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/mod.rs:412]: the corresponding PR from Thomas has been accepted so should get merged fairly soon, so if you rebase and update should be ok 👍

File: tfhe/src/shortint/engine/server_side/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/mod.rs:304]: - `generate_lookup_table` does not deal with modulus
- `generate_msg_lookup_table` applies `% msg_modulus` to the input and to the output of the function
- `generate_lookup_table_bivariate_with_factor`  which wrap `f` this way
```
let wrapped_f = |input: u64| -> u64 {
    let lhs = (input / factor_u64) % message_modulus;
    let rhs = (input % factor_u64) % message_modulus;

    f(lhs, rhs)
};
```
- `generate_lookup_table_bivariate` same as `generate_lookup_table_bivariate_with_factor` with `factor = server_key.message_modulus`

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (IceTDrinker) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs:64]: indeed for non power of two it's probably better the way you just wrote it

Comment (IceTDrinker) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs:64]: let's go with that formula then I'd say

File: tfhe/src/shortint/engine/server_side/mod.rs
Comment (tmontaigu) [reply tfhe/src/shortint/engine/server_side/mod.rs:304]: If I take the integers implementations, most of them actually play with the carry space whether its the one of the input or output, I even have a lut that ouputs a block with padding bit set.

So as long as we have `generate_lookup_table` for maximum control and `generate_lookup_table_bivariate` its fine. 

`generate_msg_lookup_table` could be used in some places to replace `generate_lookup_table` where we manually apply msg_mod to both input/output but i don't think that is very common

File: tfhe/src/shortint/engine/server_side/scalar_sub.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/scalar_sub.rs]: Done

File: tfhe/src/shortint/server_key/mod.rs
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs]: Done

File: tfhe/src/shortint/engine/server_side/mod.rs
Comment (IceTDrinker) [reply tfhe/src/shortint/engine/server_side/mod.rs:304]: all right thanks for the detailed infos, will do a last round of reviews

File: tfhe/src/high_level_api/shortints/server_key.rs
Comment (mayeul-zama) [reply tfhe/src/high_level_api/shortints/server_key.rs]: Shortints hve been removed for the HL API so I guess this solves this issue

**Review by tmontaigu — APPROVED** [review]

**Pull request merged**
